### DISCUSSION TITLE
Modify Session Object Value ?

### DISCUSSION THREAD

**Discussion by sushilbalami** [nextauthjs/next-auth#2226] [Help] [answered]
Can I modify the session object?
I am using a credentials provider so.., I am authenticating with a custom database so
although I put this obj

```
"role": "user",
"_id": "60c1f6d9abaa0e2670b06b85",
"email": "sush@gmail.com",
"createdAt": "2021-06-10T11:26:17.308Z",
"updatedAt": "2021-06-10T11:26:17.308Z",
"__v": 0 
```
this returns

```
{
  user: { name: null, email: 'sush@gmail.com', image: null },
  expires: '2021-07-11T07:08:46.365Z'
}
```
on getSession({req});

so any way to add custom key-value or custom value on session obj?

Saw this: https://next-auth.js.org/configuration/callbacks#session-callback

only name, email and image keys are parsed why so ?
`user: { name: 'sushilbalami', email: '***@gmail.com', image: null },
`
tried putting id on
`session.user.id = customId,`
but only name, email, image values are only parsed

**Comment by balazsorban44** [comment]
Check out the JWT callback as well. For increased security/privacy - by default - we only persist a subset of the user data in the JWT token.

https://next-auth.js.org/configuration/callbacks#jwt-callback

**Comment by rchampourlier** [comment, marked as answer]
Hi @sushilbalami,

I struggled a lot with this one too, trying to retrieve the `screen_name` for a Twitter authentication.

Here's my `[...nextauth].ts` file, when I finally got it working:

```typescript
import NextAuth from "next-auth";
import { signin } from "next-auth/client";
import Providers from "next-auth/providers";

export default NextAuth({
  providers: [
    Providers.Twitter({
      clientId: process.env.TWITTER_KEY,
      clientSecret: process.env.TWITTER_SECRET,
      profile(profile) {
        return {
          id: profile.id_str as string,
          name: profile.name,
          email: profile.email,
          image: (profile.profile_image_url_https as string).replace(
            /_normal\.(jpg|png|gif)$/,
            ".$1"
          ),
          screen_name: profile.screen_name,
        };
      },
    }),
  ],

  callbacks: {
    async signIn(_userOrProfile, _account, _OAuthProfile) {
      return true;
    },

    async jwt(token, user, _account, _profile, _isNewUser) {
      if (user) {
        token.screen_name = user.screen_name;
      }
      return token;
    },

    async session(session, token) {
      session.screen_name = token.screen_name;
      return session;
    },
  },
});

```

The key was to add the `screen_name` to the token and then use the `token` argument in the `session` callback to return an enriched session object.

It may be different from your case, since I'm not using a database yet, but I hope this helps!

@balazsorban44 Maybe the documentation could be improved to make this easier, especially to newcomers (I'm new to NextJS). Happy to help with a pull-request with some guidance.

Best,

**Reply by balazsorban44** [reply]
if you tell me what part of the docs is unclear, we could come up with a PR. 😊

in short, callbacks are something like this:

jwt: what will be persisted in the session cookie (in JWT format, hence the name)
session: what is returned from the jwt. usually a subset of jwt, (but you can add any static property if needed)

**Reply by rchampourlier** [reply]
Thanks for your reply. I'll try to explain what was hard for me (again, as a total newcomer to NextJS and NextAuth). I'm not exactly sure that a specific part of the doc is unclear, but making sense of everything was hard and there seems to miss some pieces. 

- I wanted to setup Twitter OAuth so I used the [Twitter provider](https://next-auth.js.org/providers/twitter)
- Following the docs, everything goes well and I'm able to make it work.
- When I get the `session` object on the client, I see that I only get a few attributes (id, name, email, image) but not the usual Twitter screen name (@something).
- This is what took me a very long time to find out. I've been able to see in the doc that for security reasons, this was filtered. But as one can see in the [Twitter provider's config](https://github.com/nextauthjs/next-auth/blob/main/src/providers/twitter.js), I should be able to get more by customizing the `provider.profile` function.
- I did that, but it didn't change anything, I was still only getting the default profile data.

Making sense that I also had to update the `jwt` callback to store the screen name in it and then, update the `session` callback to retrieve the value from the token and add it to the session, was really not easy for me.

Hope this makes sense!

**Reply by sushilbalami** [reply]
> Hi @sushilbalami,
> 
> I struggled a lot with this one too, trying to retrieve the `screen_name` for a Twitter authentication.
> 
> Here's my `[...nextauth].ts` file, when I finally got it working:
> 
> ```ts
> import NextAuth from "next-auth";
> import { signin } from "next-auth/client";
> import Providers from "next-auth/providers";
> 
> export default NextAuth({
>   providers: [
>     Providers.Twitter({
>       clientId: process.env.TWITTER_KEY,
>       clientSecret: process.env.TWITTER_SECRET,
>       profile(profile) {
>         return {
>           id: profile.id_str as string,
>           name: profile.name,
>           email: profile.email,
>           image: (profile.profile_image_url_https as string).replace(
>             /_normal\.(jpg|png|gif)$/,
>             ".$1"
>           ),
>           screen_name: profile.screen_name,
>         };
>       },
>     }),
>   ],
> 
>   callbacks: {
>     async signIn(_userOrProfile, _account, _OAuthProfile) {
>       return true;
>     },
> 
>     async jwt(token, user, _account, _profile, _isNewUser) {
>       if (user) {
>         token.screen_name = user.screen_name;
>       }
>       return token;
>     },
> 
>     async session(session, token) {
>       session.screen_name = token.screen_name;
>       return session;
>     },
>   },
> });
> ```
> 
> The key was to add the `screen_name` to the token and then use the `token` argument in the `session` callback to return an enriched session object.
> 
> It may be different from your case, since I'm not using a database yet, but I hope this helps!
> 
> @balazsorban44 Maybe the documentation could be improved to make this easier, especially to newcomers (I'm new to NextJS). Happy to help with a pull-request with some guidance.
> 
> Best,

yea cool you found out the answer ...
yea I was totally unclear on documentation
they gave me the link to callbacks I saw many times couldn't understand how to add values so.
they should have put an example saying how additional values can be added

**Reply by balazsorban44** [reply]
Please open a PR with suggestions how the docs should be better! Keep in mind we are working on this project in our free time, and although we intend to help as many as possible, we cannot satisfy everyone's needs all the time, without a little help from the community. If someone opens a PR, I'll be happy to help and review it.

**Reply by rchampourlier** [reply]
Of course @balazsorban44, thanks already for your time in the discussion and the effort put in the library. I'll try to carve out some time to propose a PR with some documentation improvements!
